File: solutions/568.maximum-vacation-days/maximum-vacation-days.py
class Solution:
    def maxVacationDays(self, flights, days):
        """
        :type flights: List[List[int]]
        :type days: List[List[int]]
        :rtype: int
        """
        N = len(flights)
        K = len(days[0])
        
        dp = [float('-inf')] * N
        dp[0] = 0
        # i week
        # j : to city
        # k : from city
        for i in range(K):
            new = [float('-inf')] * N
            for j in range(N):
                for k in range(N):
                    if (flights[k][j] == 1 or j == k):
                        new[j] = max(new[j], dp[k] + days[j][i])
                        
            dp = new
        return max(dp)

File: solutions/568.maximum-vacation-days/maximum-vacation-days.java
class Solution {
    public int maxVacationDays(int[][] flights, int[][] days) {
        // corner case


        int N = flights.length, K = days[0].length;
        int[] dp = new int[N]; // dp[i][j] is max days on week i at city j
        Arrays.fill(dp, Integer.MIN_VALUE);
        dp[0] = 0;

        for (int i = 0; i < K; i++) {
            int[] temp = new int[N];
            Arrays.fill(temp, Integer.MIN_VALUE);
            for (int j = 0; j < N; j++) {
                for (int k = 0; k < N; k++) {
                    if (j == k || flights[k][j] == 1 ) {
                        temp[j] = Math.max(temp[j], dp[k] + days[j][i]);
                    }
                }
            }
            dp = temp;
        }
        int mx = 0;
        for (int v : dp) {
            mx = Math.max(mx, v);
        }
        return mx;
    }
}